zerver/tests/test_previewable.py

```diff
@@ -0,0 +1,212 @@
+            result["pull_request"],
+            {
+                "url": "https://api.github.com/repos/zulip/zulip/pulls/22368",
+                "html_url": "https://github.com/zulip/zulip/pull/22368",
+                "diff_url": "https://github.com/zulip/zulip/pull/22368.diff",
+                "patch_url": "https://github.com/zulip/zulip/pull/22368.patch",
+                "merged_at": None,
+            },
+        )
+
+    def test_not_previewable_url(self) -> None:
+        result = self.client_post(
+            "/json/previewable", {"url": "https://github.com/zulip/xyz/zulip/issues/3"}
+        )
+        self.assert_json_error(result, msg="URL is not previewable.")
+
+    def test_not_url(self) -> None:
+        result = self.client_post(
+            "/json/previewable", {"url": "https://github.com[/zulip/xyz/zulip/issues/3"}
+        )
+        self.assert_json_error(result, msg="URL is not valid.")
+
+    @responses.activate()
+    def test_invalid_previewable_url(self) -> None:
+        response_data = (
+            {
+                "message": "Not Found",
+                "documentation_url": "https://docs.github.com/rest/reference/issues#get-an-issue",
+            },
+        )
+        api_url = "https://api.github.com/repos/zulip/zulip/issues/2"
+        responses.add(
+            responses.GET,
+            api_url,
+            json=response_data,
+            status=404,
+        )
+        response = self.client_post(
+            "/json/previewable", {"url": "https://github.com/zulip/zulip/issues/2"}
+        ).json()
+        self.assertEqual(response["msg"], "Invalid or expired URL.")
+
+    @responses.activate()
+    def test_cache(self) -> None:
+        response_data = {
+            "title": "Show titles of GitHub issues on hover",
+            "user": {"login": "brijsiyag"},
+            "state": "open",
+            "locked": False,
+            "state_reason": None,
+        }
+        responses.add(
+            responses.GET,
+            "https://api.github.com/repos/zulip/zulip/issues/19710",
+            json=response_data,
+        )
+        self.client_post(
+            "/json/previewable", {"url": "https://github.com/zulip/zulip/issues/19710"}
+        )
+        responses.remove(responses.GET, "https://api.github.com/repos/zulip/zulip/issues/19710")
+        result = self.client_post(
+            "/json/previewable", {"url": "https://github.com/zulip/zulip/issues/19710"}
+        ).json()
+        self.assertEqual(
+            result["title"],
+            "Show titles of GitHub issues on hover",
+        )
+        self.assertEqual(
+            result["user"],
+            {"login": "brijsiyag"},
+        )
+        self.assertEqual(
+            result["state"],
+            "open",
+        )
+        self.assertEqual(
+            result["locked"],
+            False,
+        )
+        self.assertEqual(
+            result["state_reason"],
+            None,
+        )
```

Comment:
These don't add much to the test, since we test that above.  What seems more interesting is that the first and second responses were both successful, and have the same body, and that the second one didn't cause an HTTP request:
```python3
        first_response = self.client_post("/json/previewable", {"url": url})
        self.assert_json_success(first_response)
        responses.assert_call_count(api_url, 1)

        second_response = self.client_post("/json/previewable", {"url": url})
        self.assert_json_success(second_response)
        responses.assert_call_count(api_url, 1)

        self.assertEqual(first_response.json(), second_response.json())